apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/observer/ShardObserver.java

```diff
@@ -59,17 +59,31 @@ public void onTrimDone(@Observes @Sync TrimData trimData) {
         boolean isShardingOff = propertiesHolder.getBooleanProperty("apl.noshardcreate", false);
         boolean shardingEnabled = currentConfig.isShardingEnabled();
         log.debug("Is sharding enabled ? : '{}' && '{}'", shardingEnabled, !isShardingOff);
+        int shardingFrequency = 1;
         if (shardingEnabled && !isShardingOff) {
+            if (!blockchainConfig.isJustUpdated()) {
+                // config didn't change from previous trim scheduling
+                shardingFrequency = currentConfig.getShardingFrequency();
+            } else {
+                // config has changed from previous trim scheduling, try to get previous 'shard frequency' value
+                shardingFrequency = blockchainConfig.getPreviousConfig().isPresent() ?
+                        blockchainConfig.getPreviousConfig().get().getShardingFrequency() // previous config
+                        : currentConfig.getShardingFrequency(); // fall back
+            }
             log.debug("Check shard conditions: ? [{}],  lastTrimBlockHeight = {}, blockchainHeight = {}"
-                    + ", shardingFrequency = {}",
-                    currentConfig.getShardingFrequency() != 0 ?
-                            lastTrimBlockHeight % currentConfig.getShardingFrequency() == 0 : "zeroDivision",
+                    + ", shardingFrequency = {} ({})",
+                    shardingFrequency != 0 ?
+                            lastTrimBlockHeight % shardingFrequency == 0 : "zeroDivision",
                     lastTrimBlockHeight, blockchainHeight,
-                    currentConfig.getShardingFrequency());
-            if (lastTrimBlockHeight != 0 && lastTrimBlockHeight % currentConfig.getShardingFrequency() == 0) {
+                    shardingFrequency, blockchainConfig.isJustUpdated());
+            if (lastTrimBlockHeight != 0 && lastTrimBlockHeight % shardingFrequency == 0) {
                 completableFuture = shardService.tryCreateShardAsync(lastTrimBlockHeight, blockchainHeight);
             } else {
-                log.debug("No attempt to create new shard at height '{}'", blockchainHeight);
+                log.debug("No attempt to create new shard at height '{}' (because lastTrimHeight={}), ({})",
+                        blockchainHeight, lastTrimBlockHeight, blockchainConfig.isJustUpdated());
+            }
+            if (blockchainConfig.isJustUpdated()) {
```

Comment:
I think that its not the right place for it

Reply:
It works fine and consistent. What is trouble here?

Reply:
Better to use BlockchainConfigUpdater

Reply:
BlockchainConfigUpdater doesn't help here.

Reply:
Take a look at BlockchainConfigUpdater.onBlockAccepted. Current impl is a violation of single responsibility principle.

Reply:
Config keeps his own state if it was updated. I don't see any violation,